kernel/kernel/fs/pipe.cpp

```diff
@@ -852,6 +853,79 @@ ssize_t pipe::read_iter(iovec_iter *iter, unsigned int flags)
 
 }
 
+ssize_t pipe::write_iter(iovec_iter *iter, int flags)
+{
+    bool is_atomic_write = iter->bytes <= PIPE_BUF;
+    ssize_t ret = 0;
+
+    scoped_mutex g{pipe_lock};
+
+    bool wasempty = !can_read();
+
+    while (!iter->empty())
+    {
+        if (reader_count == 0)
+        {
+            CALL_KUNIT_MOCKABLE(kernel_raise_signal, SIGPIPE, get_current_process(), 0, nullptr);
+
+            if (!ret)
+                ret = -EPIPE;
+            break;
+        }
+
+        const auto avail = available_space();
+
+        bool may_write = avail > 0 && !(avail < iter->bytes && is_atomic_write);
+
+        if(!may_write)
+        {
+            if (ret != 0)
+            {
+                /* now that we're blocking, might as well signal readers */
+                wake_all(&read_queue);
+            }
+
+            if (flags & O_NONBLOCK)
+            {
+                if (!ret)
+                    ret = -EAGAIN;
+                break;
+            }
+
+            if (wait_for_event_mutex_interruptible(
+                    &write_queue,
+                    (is_atomic_write && available_space() >= (iter->bytes - ret)) || !is_full() ||
```

Comment:
Hmm. I know that the ->write code doesn't handle this. But I think this could be slightly simplified to

`(is_atomic_write && available_space() >= iter->bytes) || !is_full() ||`

because is_atomic_write means (hopefully) that we're not going to do a partial write, and ret is used to know how much we've written from to the pipe (for partial writes). I think this is correct, but I haven't looked at this code for a while. Thoughts?

Reply:
Yes, this should be correct, since iter->bytes shows amount of unwritten bytes